File: src/Queue.py
from typing import Optional


class Node:
    def __init__(self, value):
        self.value = value
        self.next: Optional["Node"] = None


class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def EnQueue(self, item):
        node = Node(item)
        self.length += 1

        if self.tail is None:
            self.head = self.tail = node
            return
        self.tail.next = node
        self.tail = node

    # Method to remove an item from the queue
    def DeQueue(self):
        if self.head is None:
            return

        self.length -= 1
        head = self.head
        self.head = self.head.next

        if self.head is None:
            self.tail = None

File: src/Queue.ts
type Node<T> = {
  value: T;
  nextNode?: Node<T>;
};

export class Queue<T> {
  public length: number;
  public head?: Node<T>;
  public tail?: Node<T>;

  constructor() {
    this.head = this.tail = undefined;
    this.length = 0;
  }

  enqueue(item: T) {
    const node = { value: item } as Node<T>;
    this.length++;

    if (!this.tail) {
      this.head = this.tail = node;
      return;
    }

    this.tail.nextNode = node;
    this.tail = node;
  }

  deque() {
    if (!this.head) {
      return undefined;
    }
    this.length--;
    const head = this.head;
    this.head = this.head.nextNode;

    if (this.length === 0) {
      this.tail = undefined;
    }

    return head.value;
  }

  peek() {
    return this.head?.value;
  }
}
